metric_tank/aggmetric.go

```diff
@@ -247,33 +253,117 @@ func (a *AggMetric) addAggregators(ts uint32, val float64) {
 	}
 }
 
-func (a *AggMetric) Persist(c *Chunk) {
-	log.Debug("AggMetric %s Persist(): starting to save %v", a.Key, c)
-	data := c.Series.Bytes()
-	chunkSizeAtSave.Value(int64(len(data)))
+func (a *AggMetric) persist(pos int) {
+	chunk := a.Chunks[pos]
+	chunk.Finish()
+	if *dryRun {
+		chunk.Saved = true
+		return
+	}
 
-	version := FormatStandardGoTsz
-	buf := new(bytes.Buffer)
-	err := binary.Write(buf, binary.LittleEndian, uint8(version))
-	if err != nil {
-		// TODO
+	// this will block when the write queue fills up.
+	log.Debug("sending chunk to write queue")
+	timer := time.NewTicker(2 * time.Second)
+WAIT:
+	for {
+		select {
+		case a.writeQueue <- chunk:
+			log.Debug("chunk in write queue: length: %d", len(a.writeQueue))
+			break WAIT
+		case <-timer.C:
+			log.Warn("%s:%d blocked pushing to writeQueue.", a.Key, chunk.T0)
+		}
 	}
-	_, err = buf.Write(data)
-	if err != nil {
-		// TODO
+	timer.Stop()
+
+	if !a.activeWrite {
+		log.Debug("starting persist goroutine.")
+		a.activeWrite = true
+		// asynchronously write data to cassandra.
+		go func() {
+			for {
+				select {
+				case c := <-a.writeQueue:
```

Comment:
This always runs at least once.  this goroutine can only be started or stopped by a thread holding the lock.  The comments at line 284 and 340 provide details on this.